File: cpp/trapRain.cpp
#include <vector>
#include <algorithm>

using namespace std;

class Solution {
public:
    int trap(vector<int>& h) {
        int l = 0;
        int r = h.size() - 1;
        int lmax = -1;
        int rmax = -1;
        int total = 0;

        while(l<r){
            lmax=max(lmax,h[l]);
            rmax=max(rmax,h[r]);
            total += (lmax<rmax)? lmax-h[l++] : rmax-h[r--];
        }
        return total;
    }
};

File: py/trapRain.py
class Solution:
    def trap(self, height: List[int]) -> int:
        l = 0
        r = len(height) - 1
        total = 0
        lmax = -1
        rmax = -1

        while l < r:
            lmax = max(lmax, height[l])
            rmax = max(rmax, height[r])

            if lmax < rmax:
                total += lmax - height[l]
                l += 1
            else:
                total += rmax - height[r]
                r -= 1

        return total
        